text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+
+**FPC Validator:** This validation complements the Peer validation logic by validating transactions produced by an FPC Chaincode.
+It does this by checking that each transaction has been cryptographically signed (endorsed) by a Chaincode Enclave that is listed in the FPC Registry and is authorized for that chaincode.
+
+**FPC Client SDK extension:** This extends the Fabric Client SDKs with extra functionality that allows users to write end-to-end secure FPC-based applications.
+In particular, the FPC Client SDK extension will provide these core functions: 1) FPC transaction proposal creation (including transparent encryption of arguments) 2) FPC transaction proposal response validation and decryption of the result.
+The encryption and decryption is performed by the Client SDK "under the covers" without requiring any special action by the users, i.e.,
+the users still use normal `invoke`/`query` functions to issue FPC transaction invocations.
+For the MVP, we intend to support the peer CLI only. Extended support for the NodeSDK (or Go SDK) will be future work.
+
+**FPC SDK:** The FPC SDK contains a set of tools to write and build chaincodes.
+In particular, the SDK contains a build-chain to link a chaincode against our FPC Shim and produce a deployment artifact that can be executed by the FPC runtime.
+
+**FPC Runtime:** The FPC runtime comprises all components that are required to deploy and run an FPC chaincode.
+
+
+## FPC chaincode deployment
+
+The deployment of a FPC chaincode consists of multiple steps; it begins with the standard Fabric 2.0 Lifecycle mechanism to produce a chaincode definition that all participants agree upon. FPC requires adding some additional metadata to the chaincode definition, in particular a cryptographic hash (called the MRENCLAVE in the initial version of FPC) that identifies the code running inside the Chaincode Enclave. The FPC SDK provides tools to generate this MRENCLAVE in a way that enables all parties to screen the code upfront and produce this hash. Note that this is crucial for the security provided by the TEE and is part of the attestation capabilities.
+For more details on Intel SGX remote attestation we refer to the vendor [documentation](https://software.intel.com/content/www/us/en/develop/topics/software-guard-extensions/attestation-services.html).
+
+Once the chaincode definition for an FPC chaincode has been approved by the consortium, a Chaincode Enclave is created by calling `peer lifecycle chaincode createenclave`. This command triggers the creation of an enclave and registers it at the FPC Registry.  The registration includes the attestation of the Chaincode Enclave.
```

Comment:
I think we need to find an alternative way of creating the enclave ... adding a command to the lifecycle seems like the wrong approach here.

Reply:
If we had a CLI that is extendable via plugins (what's the status of https://github.com/hyperledger/fabric-cli ? Is it ongoing work? abandoned?) then it'd make sense to implement the support there.

Reply:
Conceptually, the enclave creation is part of the chaincode lifecycle, hence our proposal to put it into the corresponding namespace of closely related commands in the peer cli. Implementation-wise, we currently handle that we a simple script which wraps and extends for FPC the normal command. The actual protocol message would initially be layered on top of a normal query flow.

To better understand your concern, is your concern the particular namespace used or that this command is exposed to admins at all? Or something else?

Changing namespace of course would be trivial. For the short-term MVP solution with a designated peer, we also could simply hide the enclave creation and trigger it "under the cover".  It is a bit less clear whether hiding enclave interaction still easily works if we go to more advanced endorsement policy models which involve additional key distribution protocol flows. Hence, we were conservative and thought we propose something which works for sure also longer term. That said we of course will try handle additional requirements and other suggestions.